proc/sort.go

```diff
@@ -1,55 +1,189 @@
 package proc
 
 import (
+	"bufio"
+	"container/heap"
 	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
 
 	"github.com/brimsec/zq/ast"
 	"github.com/brimsec/zq/expr"
 	"github.com/brimsec/zq/pkg/nano"
 	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio/bzngio"
 	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
 )
 
-type ErrSortLimitReached int
-
-func (e ErrSortLimitReached) Error() string {
-	return fmt.Sprintf("sort limit hit (%d)", e)
-}
+// sortMemMaxBytes specifies the maximum amount of memory that each sort proc
+// will consume.
+var SortMemMaxBytes = 128 * 1024 * 1024
 
 type Sort struct {
 	Base
 	dir        int
-	limit      int
 	nullsFirst bool
 	fields     []ast.FieldExpr
-	resolvers  []expr.FieldExprResolver
-	out        []*zng.Record
-}
 
-// defaultSortLimit is the default limit of the number of records that
-// sort will process and, otherwise, return an error if this limit is exceeded.
-// The value can be overridden by setting the limit param on the SortProc.
-const defaultSortLimit = 1000000
+	fieldResolvers     []expr.FieldExprResolver
+	once               sync.Once
+	resultCh           chan Result
+	sortFn             expr.SortFn
+	unseenFieldTracker *unseenFieldTracker
+}
 
 func CompileSortProc(c *Context, parent Proc, node *ast.SortProc) (*Sort, error) {
-	limit := node.Limit
-	if limit == 0 {
-		limit = defaultSortLimit
-	}
-	resolvers, err := expr.CompileFieldExprs(node.Fields)
+	fieldResolvers, err := expr.CompileFieldExprs(node.Fields)
 	if err != nil {
 		return nil, err
 	}
 	return &Sort{
-		Base:       Base{Context: c, Parent: parent},
-		dir:        node.SortDir,
-		limit:      limit,
-		nullsFirst: node.NullsFirst,
-		fields:     node.Fields,
-		resolvers:  resolvers,
+		Base:               Base{Context: c, Parent: parent},
+		dir:                node.SortDir,
+		nullsFirst:         node.NullsFirst,
+		fields:             node.Fields,
+		fieldResolvers:     fieldResolvers,
+		resultCh:           make(chan Result),
+		unseenFieldTracker: newUnseenFieldTracker(node.Fields, fieldResolvers),
 	}, nil
 }
 
+func (s *Sort) Pull() (zbuf.Batch, error) {
+	s.once.Do(func() { go s.sortLoop() })
+	r := <-s.resultCh
+	return r.Batch, r.Err
+}
+
+func (s *Sort) sortLoop() {
+	defer close(s.resultCh)
+	firstRunRecs, eof, err := s.recordsForOneRun()
+	if err != nil || len(firstRunRecs) == 0 {
+		s.sendResult(nil, err)
+		return
+	}
+	s.setSortFn(firstRunRecs[0])
```

Comment:
Rather than have to do separate logic for the first in memory "run", I think it would be cool to create something like `zbuf.Peeker` and do something like:
```golang
rec, err := s.peeker.Peek(1)
if err != nil {
    return err
}
s.setSortFn(rec)
```

Reply:
@mattnibs: Is part of the idea here that the logic for keeping the first run in memory would live in FileMergeSort, so we'd need another way to get at the first record?